### PR TITLE
JAMES-2644 Fix the Content-Type parsing fragility

### PR DESCRIPTION
**Pull request by blackheaven**

### CODE REVIEW

File: server/container/core/src/test/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxyTest.java [hunk 1]
```diff
@@ -245,6 +246,20 @@ public void testMessageCloningViaCoW3() throws Exception {
         LifecycleUtil.dispose(mm);
     }
 
+    @Test
+    @Disabled("JAMES-2644")
+    public void testMessageWithWrongContentTypeShouldNotThrow() throws Exception {
+        ArrayList<MailAddress> r = new ArrayList<>();
+        r.add(new MailAddress("recipient@test.com"));
```
Comment (trantienduchn) [inline server/container/core/src/test/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxyTest.java]: why did you named it to `r`? And you could very well use `ImmutableList.of()`

**Comment by blackheaven** [comment]
The aim of this PR is to provide a suitable test which will be executed when the library will be replaced

**Comment by rouazana** [comment]
> The aim of this PR is to provide a suitable test which will be executed when the library will be replaced

Which library? That's not clear neither from this PR, nor from the ticket you mention. BTW did you check this library cannot be configured to accept such malformed data?

**Comment by rouazana** [comment]
> It's javax.mail and no option go in the parsing method.

no, but javax.mail handles parting options through properties, see for example https://docs.oracle.com/javaee/6/api/javax/mail/internet/package-summary.html#package_description

**Comment by blackheaven** [comment]
I'll have a look, thanks

**Comment by chibenwa** [comment]
test this please

File: server/container/core/src/main/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxy.java [hunk 2]
```diff
@@ -46,6 +47,15 @@
  * automatically cloning it (if shared) when a write operation is invoked.
  */
 public class MimeMessageCopyOnWriteProxy extends MimeMessage implements Disposable {
+    public static class ContentTypeCleaner {
+        public static String cleanContentType(MimePart mimePart, String contentType) {
+            return null;
```
Comment (trantienduchn) [inline server/container/core/src/main/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxy.java]: is content-type is save to remove?

Comment (blackheaven) [reply server/container/core/src/main/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxy.java]: This method is called when Content-Type can not be parsed and when Content-Type is `null`, it is ignored, so I think it pretty safe.

Comment (trantienduchn) [reply server/container/core/src/main/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxy.java]: We can return a string has been removed all slashes. WDYT?

Comment (blackheaven) [reply server/container/core/src/main/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxy.java]: Basically, it's an issue created due to production failure and there are already no slashes.

At this point, we are unable to extract a valid `Content-Type`, making up another one looks dangerous.

@rouazana Maybe I can add a log wan/error, WDYT?

Comment (rouazana) [reply server/container/core/src/main/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxy.java]: @trantienduchn the issue is not having a `/`, it's that there is a `;` while a `/` is expected. The content-type here seems to be really screwed. It should not be "string" but probably "application/pdf". But we have no way to recover it.
Ok for me to add a warn log.

File: server/container/core/src/main/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxy.java [hunk 3]
```diff
@@ -41,14 +42,28 @@
...
  * This object wraps a "possibly shared" MimeMessage tracking copies and
  * automatically cloning it (if shared) when a write operation is invoked.
  */
 public class MimeMessageCopyOnWriteProxy extends MimeMessage implements Disposable {
     public static class ContentTypeCleaner {
+        private static final Logger LOGGER = LoggerFactory.getLogger(ContentTypeCleaner.class);
+        private static final Pattern REGEX = Pattern.compile("^[\\w\\-]+/[\\w\\-]+");
+
         public static String cleanContentType(MimePart mimePart, String contentType) {
+            if (contentType == null) {
```
Comment (trantienduchn) [inline server/container/core/src/main/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxy.java]: is this method tested?

**Comment by trantienduchn** [comment]
```
[d6efc6a6473538997bf01f0743ce675eae5709a7] [ERROR] Errors: 
[d6efc6a6473538997bf01f0743ce675eae5709a7] [ERROR]   MimeMessageCopyOnWriteProxyTest.testMessageWithWrongContentTypeShouldNotThrow:257 » Parse
[d6efc6a6473538997bf01f0743ce675eae5709a7] [INFO] 
[d6efc6a6473538997bf01f0743ce675eae5709a7] [ERROR] Tests run: 178, Failures: 0, Errors: 1, Skipped: 0
```
seems related

File: server/container/core/src/main/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxy.java [hunk 4]
```diff
@@ -41,14 +42,28 @@
...
+                return null;
+            }
+
+            if (REGEX.matcher(contentType).find()) {
```
Comment (rouazana) [inline server/container/core/src/main/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxy.java]: that's not exactly the definition in the rfc, but it seems acceptable to me

File: server/container/core/src/main/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxy.java [hunk 5]
```diff
@@ -41,14 +42,28 @@
...
+                return contentType;
+            }
+
+            LOGGER.warn("Can not parse: " + contentType);
```
Comment (rouazana) [inline server/container/core/src/main/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxy.java]: maybe a bit more context could be welcomed? Like `Can not parse content-type: `?

**Comment by rouazana** [comment]
btw you should rename your commit now ^^

File: server/container/core/src/main/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxy.java [hunk 6]
```diff
@@ -36,16 +37,40 @@
...
+            return null;
+        }
+    }
+
+    static {
+        System.setProperty("mail.mime.contenttypehandler", ContentTypeCleaner.class.getName());
```
Comment (mbaechler) [inline server/container/core/src/main/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxy.java]: Well, that's one more argument against javax.mail usage. I add it to my personal list for the day we'll argue whether we should drop that lib.

File: server/container/core/src/test/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxyTest.java [hunk 7]
```diff
@@ -245,6 +247,18 @@ public void testMessageCloningViaCoW3() throws Exception {
...
     }
 
+    @Test
+    public void testMessageWithWrongContentTypeShouldNotThrow() throws Exception {
+        ImmutableList<MailAddress> recipients = ImmutableList.of(new MailAddress("recipient@test.com"));
+        MimeMessageCopyOnWriteProxy messageFromSources = (MimeMessageCopyOnWriteProxy) getMessageFromSources(
+                content + sep + body);
+        MailImpl mail = new MailImpl("test", new MailAddress("test@test.com"), recipients, messageFromSources);
+        mail.getMessage().addHeader("Content-Type", "file;name=\"malformed.pdf\"");
+        mail.getMessage().saveChanges();
```
Comment (mbaechler) [inline server/container/core/src/test/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxyTest.java]: As far as I understand, the bug is not specific to MimeMessageCopyOnWriteProxy but is common to MimeMessage, don't you think ?

Comment (chibenwa) [reply server/container/core/src/test/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxyTest.java]: Maybe we can delegate that responsibility in a Guice ConfigurationPerformer to set the static properties for Javax.mail system wide?

Comment (blackheaven) [reply server/container/core/src/test/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxyTest.java]: I like that, thanks

File: server/container/core/src/main/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxy.java [hunk 8]
```diff
@@ -46,7 +46,6 @@
  * automatically cloning it (if shared) when a write operation is invoked.
  */
 public class MimeMessageCopyOnWriteProxy extends MimeMessage implements Disposable {
-
```
Comment (rouazana) [inline server/container/core/src/main/java/org/apache/james/server/core/MimeMessageCopyOnWriteProxy.java]: unrelated change

File: server/container/core/pom.xml [hunk 9]
```diff
@@ -128,6 +128,10 @@
             <artifactId>JUnitParams</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.google.inject</groupId>
+            <artifactId>guice</artifactId>
```
Comment (chibenwa) [inline server/container/core/pom.xml]: Euh no I don't want that dependency everywhere!

What we could have is a static `ContentTypeCleaner.initialize()` method here, and define a guice module `JavaxModule` in `server/container/guice/guice-common` that would be then bound by all products.

File: server/container/core/src/main/java/org/apache/james/server/core/MimeMessageModule.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+public class MimeMessageModule extends AbstractModule {
+
+    @Override
+    protected void configure() {
+        System.setProperty("mail.mime.contenttypehandler", ContentTypeCleaner.class.getName());
+    }
+
+    public static class ContentTypeCleaner {
+        private static final Logger LOGGER = LoggerFactory.getLogger(ContentTypeCleaner.class);
+        private static final Pattern REGEX = Pattern.compile("^[\\w\\-]+/[\\w\\-]+");
```
Comment (chibenwa) [inline server/container/core/src/main/java/org/apache/james/server/core/MimeMessageModule.java]: I am a REGEX noob. Could you explain a bit more what this regex is doing?

Comment (blackheaven) [reply server/container/core/src/main/java/org/apache/james/server/core/MimeMessageModule.java]: It looks for a string beginning by a bunch alphanumeric characters or '-' or '_' followed by a '/', then a bunch alphanumeric characters or '-' or '_'.

File: server/container/guice/guice-common/src/main/java/org/apache/james/modules/CommonServicesModule.java [hunk 11]
```diff
@@ -61,6 +61,7 @@ protected void configure() {
         install(new DropWizardMetricsModule());
         install(new TaskManagerModule());
         install(new CleanupTaskModule());
+        install(new MimeMessageModule());
```
Comment (chibenwa) [inline server/container/guice/guice-common/src/main/java/org/apache/james/modules/CommonServicesModule.java]: I would name it `JavaxMailConfigurationModule`

File: server/container/core/src/main/java/org/apache/james/server/core/ContentTypeCleaner.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ContentTypeCleaner {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ContentTypeCleaner.class);
+    private static final Pattern REGEX = Pattern.compile("^[\\w\\-]+/[\\w\\-]+");
+
+    public static void initialize() {
+        System.setProperty("mail.mime.contenttypehandler", ContentTypeCleaner.class.getName());
```
Comment (aduprat) [inline server/container/core/src/main/java/org/apache/james/server/core/ContentTypeCleaner.java]: You can extract the key in  a constant.

File: server/container/core/src/test/java/org/apache/james/server/core/ContentTypeCleanerTest.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+ ****************************************************************/
+
+package org.apache.james.server.core;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.junit.jupiter.api.Test;
+
+public class ContentTypeCleanerTest {
+    @Test
```
Comment (aduprat) [inline server/container/core/src/test/java/org/apache/james/server/core/ContentTypeCleanerTest.java]: You can add a test on an empty content type (`contentType = ""`)

File: server/container/core/src/main/java/org/apache/james/server/core/ContentTypeCleaner.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    public static String cleanContentType(MimePart mimePart, String contentType) {
+        if (contentType == null) {
```
Comment (aduprat) [inline server/container/core/src/main/java/org/apache/james/server/core/ContentTypeCleaner.java]: `Strings.isNullOrEmpty` to deal with empty case.

File: server/container/core/pom.xml [hunk 15]
```diff
@@ -141,7 +141,7 @@
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-surefire-plugin</artifactId>
                 <configuration>
-                    <reuseForks>true</reuseForks>
+                    <reuseForks>false</reuseForks>
```
Comment (rouazana) [inline server/container/core/pom.xml]: is it related?

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/SmtpContentTypeTest.java [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+
+    @Test
+    public void userShouldBeAbleToReceiveMessagesWithGoodContentType() throws Exception {
+        createJamesServer(SmtpConfiguration.builder()
+            .requireAuthentication()
+            .withAutorizedAddresses("172.0.0.0/8"));
+
+        messageSender.connect(LOCALHOST_IP, jamesServer.getProbe(SmtpGuiceProbe.class).getSmtpPort())
+            .authenticate(FROM, PASSWORD)
+            .sendMessage(mailWithContentType("test/plain;"));
```
Comment (rouazana) [inline server/mailet/integration-testing/src/test/java/org/apache/james/smtp/SmtpContentTypeTest.java]: s/test/text/ ?

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/SmtpContentTypeTest.java [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+
+        awaitAtMostOneMinute
+            .untilAsserted(() -> fakeSmtp.assertEmailReceived(response -> response
+                .body("", hasSize(1))
+                .body("[0].from", equalTo(FROM))
+                .body("[0].subject", equalTo(SUBJECT))));
```
Comment (rouazana) [inline server/mailet/integration-testing/src/test/java/org/apache/james/smtp/SmtpContentTypeTest.java]: you don't check you get the content type?

File: server/container/core/pom.xml (hunk 15 above)
Comment (blackheaven) [reply server/container/core/pom.xml]: Yes because the environment variable is checked once and then cached, so, reusing the fork mean that setting the environment variable have no effect.

**Comment by chibenwa** [comment]
Merged
